Lua 5.2+ compatbility

mcepl opened the pull request:
This is compatibility layer, which makes the code run both on Lua 5.1 (which is the default for Neovim, and it is what LuaJIT provides) and Lua 5.2+.

As mentioned in https://github.com/neovim/neovim/issues/7879#issuecomment-441732865

mcepl committed:
test: Lua 5.2/5.3 compat #9280

Make the code run both on Lua 5.1 (which is the default for Neovim, and
is what LuaJIT provides) and Lua 5.2+.

jamessan commented:
Should we have `compat` modules to contain these types of things?  For example, `test/compat.lua` exports `unpack`, appropriately defined and other modules use `compat.unpack`.

Do we need a lint check to avoid adding new uses of unpack which aren't compatible?

bfredl commented:
What about lua plugin code that is written for nvim+lua5.1 (the officially supported build)? What should we tell the user when such plugins  break on this separate nvim+5.2 config? What if someone develops a plugin on nvim+5.2, not knowing the build has been altered, and it break for someone using the official lua5.1/luajit config?

An option is just to make it clearer that distros are expected to use lua 5.1 or luajit. We could could make the buildscripts fail if lua 5.2/5.3 is used. Distros can statically link the correct version with nvim, nvim can't hardly be the only software that expects a given lua 5.x version, given all the incompatibilities.

bfredl commented:
Though to be less strict, we can still allow external build/generator/test code _itself_ to be run by the distros lua. The issue is with the expectations for people writing runtime plugins.

mcepl commented:
> An option is just to make it clearer that distros are expected to use lua 5.1 or luajit.

I think it is a bad idea. It would make Neovim hard to build for distributions like Fedora which has Lua 5.3 as the default Lua and not all packages for lua 5.1 are available (and I am neither Fedora user nor maintainer any more).

bfredl commented:
> not all packages for lua 5.1 are available

Again, we could support 5.2/5.3 for the _build_ process, which has the all the dependencies for extra binary lua packages (lpeg etc). But for _runtime_ support only `lua-5.1` itself is needed, nothing extra. It would be nice if a plugin developer need not test their code on multiple different configurations.

jamessan commented:
The unfortunate part of the Lua ecosystem is that the different Lua versions aren't compatible.  This forces distributors to choose _a_ Lua version when they need to compile a program.

I'm not sure about Fedora, but I know that Debian's Lua packaging makes it trivial to build modules for multiple Lua versions.  However, that doesn't apply to binaries (like vim and nvim).  Those need to choose a specific version of Lua, which will then restrict what modules can be used if module upstreams don't target all Lua versions.

This problem already exists with Vim since Vim explicitly supports being built against any of the current Lua versions.  I think the number of Lua plugins written for Vim is tiny, though, whereas nvim is specifically treating Lua as a first class citizen.  The expectation is that there will be a lot more Lua written for nvim.

If we allow nvim to be built against any Lua version, then plugin authors need to account for that.  However, if we don't allow that, then we potentially make it harder for people to use modules they want.

hkupty commented:
As a plugin developer, one of the things I expect is solid foundations to rely on, such as always being 5.1 on runtime. Otherwise I'd need to have multiple different abstraction layers, which is absolutely suboptimal.

As a fedora user, I solved the issue of having no default lua-5.1 by using supported out-of-the-tree packages to provide me lua-5.1 and luarocks-5.1.

That's a small hurdle, but neovim itself was already packaged correctly, so that was only required for me to write and test plugins.

mcepl commented:
> As a fedora user, I solved the issue of having no default lua-5.1 by using supported out-of-the-tree packages to provide me lua-5.1 and luarocks-5.1.

Except you really cannot have out-of-the-tree packages for the official neovim, right?

bfredl commented:
> Except you really cannot have out-of-the-tree packages for the official neovim, right?

Why not? nvim with lua 5.1 can load any external `.so` lua lib that is compiled for lua 5.1.

hkupty commented:
@mcepl

> That's a small hurdle, but neovim itself was already packaged correctly.

I didn't have to install lua-5.1 to have neovim. It was already distributed with Lua correctly embedded (IIRC luajit).

Don't mix my requirement as a plugin developer with neovim install/usage experience.

justinmk commented:
as @jamessan mentioned there is a lint issue. https://github.com/neovim/neovim/pull/9280#issuecomment-442111217
Then we can merge this.

> Should we have compat modules to contain these types of things? For example, test/compat.lua exports unpack, appropriately defined and other modules use compat.unpack.

For simple cases that seems reasonable. And maybe for our use case we won't run into many 5.2/5.3-incompatible situations. For example there's no need for Nvim-Lua code to use `setfenv` nor `os.execute`.

There is also [lua-compat-5.2](https://github.com/keplerproject/lua-compat-5.2) or [lua-compat-5.3](https://github.com/keplerproject/lua-compat-5.3) but those seem risky and may invite more problems than solutions.

mcepl commented:
> Lint _does_ fail currently:
> 
> ```
> Checking /home/travis/build/neovim/neovim/test/functional/api/server_requests_spec.lua 1 warning
>     /home/travis/build/neovim/neovim/test/functional/api/server_requests_spec.lua:5:16: accessing undefined field unpack of global table
> ```

I don't understand this. It is exactly the main construct we use, right?

```lua
local unpack = table.unpack or unpack
```

I can understand that it may be problematic for a lint because it always expects to fail on at least one branch of the ``or`` expression. However, why lint complains only here not with *every other* use of this construct?

bfredl commented:
@mcepl Maybe the linter simply avoids repeating complaints of the same undefined name `table.unpack` ?

mcepl commented:
> @mcepl Maybe the linter simply avoids repeating complaints of the same undefined name `table.unpack` ?

What I meant is that linters are not fountains of all wisdom, and in this case I don't see anything particularly wrong with what I did.

bfredl commented:
I just answered the direct question "why lint complains only here not with every other use of this construct?". I didn't say anything about wisdom, it is just a simple rule to avoid excessive output (but e g gcc will say "further references omitted", so it is clearer). Somewhere it is configured what global names are allowed (nvim already adds some), one would need to add `table.unpack` there.

bfredl commented:
The only whitelisted global seems to be `vim` on line 691 in `CMakeLists.txt`. Maybe one could turn it to a proper cmake list, or separate file.

mcepl commented:
@bfredl that wasn't against you. I was reacting more to @justinmk who seemed to indicate (in https://github.com/neovim/neovim/pull/9280#issuecomment-442667821), that this cannot be merged unless the lint issues are resolved.

bfredl commented:
>  @justinmk who seemed to indicate (in #9280 (comment)), that this cannot be merged unless the lint issues are resolved.

Well, linter issues have to be resolved, in the sense that the build is not red. If the linter is wrong, the proper resolution is to reconfigure the linter, I don't think anyone is against that.

justinmk committed:
runtime/lua/vim/compat.lua

ref #9280
Introduce the `vim.compat` module, to help environments with system Lua
5.2+ run the build/tests. Include the module implicitly in all tests.

ref #8677
legacy `vim` module:
    beep
    buffer
    command
    dict
    eval
    firstline
    lastline
    line
    list
    open
    type
    window

justinmk commented:
@mcepl I see that `man.lua` was modified as well, but it would run using Nvim's builtin Lua, not system Lua--that is, runtime, not build-time. Is Nvim being _linked to_ Lua 5.2? That's a bit surprising since I thought the 5.1/5.2 C APIs were incompatible.

justinmk commented on runtime/lua/man.lua:
```diff
@@ -1,3 +1,5 @@
+require('vim.compat')
```
Not sure if we want to do this implicitly instead. But let's discuss that later.

justinmk commented on cmake/RunTests.cmake:
```diff
@@ -38,7 +38,10 @@ set(ENV{SYSTEM_NAME} ${SYSTEM_NAME})
     --lua=${LUA_PRG} --lazy --helper=${TEST_DIR}/${TEST_TYPE}/preload.lua
-    --lpath=${BUILD_DIR}/?.lua --lpath=?.lua ${TEST_PATH}
+    --lpath=${BUILD_DIR}/?.lua
+    --lpath=${WORKING_DIR}/runtime/lua/?.lua
```
This is a try at sharing code between tests/runtime/core. Maybe we will do it differently, I will have a followup PR for further discussion, after this one. 

Related: #8677  cc @KillTheMule

KillTheMule replied:
I don't understand makefiles, so just a thought: Is `?` recursive, so it also catches `runtime/lua/plugin/init.lua`? Also, does the test runner have access to `vim.api`? If so, great boon to writing tests!

justinmk replied:
yes it's recursive.

> Also, does the test runner have access to vim.api? If so, great boon to writing tests!

`vim.api` is built-into Nvim so that would not be "shared". The test utilities could spawn `nvim` to use its features. We probably should do that, to reduce some of the code in `*/helpers.lua`.

justinmk commented:
@mcepl I pushed to your branch. Could you confirm if this work? Then will merge.

mcepl commented:
> @mcepl Is Nvim being _linked to_ Lua 5.2? That's a bit surprising since I thought the 5.1/5.2 C APIs were incompatible.

[Fedora Rawhide package](http://widehat.opensuse.org/opensuse/repositories/home:/mcepl:/neovim/Fedora_Rawhide/x86_64/neovim-0.3.2~git.1543373357.0d1e5ec1b-3.2.x86_64.rpm) shows this:

```
matej@stitny: tmp$ rpm -qRp neovim*.rpm
/bin/sh
/bin/sh
/usr/bin/sh
config(neovim) = 0.3.2~git.1543373357.0d1e5ec1b-3.2
gperf
libc.so.6()(64bit)
libc.so.6(GLIBC_2.14)(64bit)
libc.so.6(GLIBC_2.2.5)(64bit)
libc.so.6(GLIBC_2.3)(64bit)
libc.so.6(GLIBC_2.4)(64bit)
libdl.so.2()(64bit)
libjemalloc.so.2()(64bit)
liblua-5.3.so()(64bit)
libm.so.6()(64bit)
libm.so.6(GLIBC_2.2.5)(64bit)
libm.so.6(GLIBC_2.29)(64bit)
libmsgpackc.so.2()(64bit)
libnsl.so.2()(64bit)
libpthread.so.0()(64bit)
libpthread.so.0(GLIBC_2.2.5)(64bit)
librt.so.1()(64bit)
libtermkey.so.1()(64bit)
libunibilium.so.4()(64bit)
libutil.so.1()(64bit)
libutil.so.1(GLIBC_2.2.5)(64bit)
libuv.so.1()(64bit)
libvterm.so.0()(64bit)
python3-nvim
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(FileDigests) <= 4.6.0-1
rpmlib(PartialHardlinkSets) <= 4.0.4-1
rpmlib(PayloadFilesHavePrefix) <= 4.0-1
rpmlib(PayloadIsXz) <= 5.2-1
rpmlib(TildeInVersions) <= 4.10.0-1
rtld(GNU_HASH)
xsel
matej@stitny: tmp$
```
And when I pull out a `/usr/bin/nvim` from that package I get:
```
matej@stitny: tmp$ ldd nvim
./nvim: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by ./nvim)
	linux-vdso.so.1 (0x00007ffc88315000)
	libuv.so.1 => /usr/lib64/libuv.so.1 (0x00007f529d975000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f529d5a3000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f529d384000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f529d180000)
	libnsl.so.2 => /usr/lib64/libnsl.so.2 (0x00007f529cf67000)
	libmsgpackc.so.2 => /usr/lib64/libmsgpackc.so.2 (0x00007f529d96b000)
	libvterm.so.0 => /usr/lib64/libvterm.so.0 (0x00007f529d955000)
	libtermkey.so.1 => /usr/lib64/libtermkey.so.1 (0x00007f529d949000)
	libunibilium.so.4 => /usr/lib64/libunibilium.so.4 (0x00007f529d934000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f529cbd4000)
	libutil.so.1 => /lib64/libutil.so.1 (0x00007f529c9d1000)
	liblua-5.3.so => not found
	libjemalloc.so.2 => /usr/lib64/libjemalloc.so.2 (0x00007f529c712000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f529c352000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f529d7ab000)
	libtirpc.so.3 => /lib64/libtirpc.so.3 (0x00007f529c120000)
	libtinfo.so.6 => /lib64/libtinfo.so.6 (0x00007f529d901000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f529d8e7000)
	libgssapi_krb5.so.2 => /usr/lib64/libgssapi_krb5.so.2 (0x00007f529d897000)
	libkrb5.so.3 => /usr/lib64/libkrb5.so.3 (0x00007f529c041000)
	libk5crypto.so.3 => /usr/lib64/libk5crypto.so.3 (0x00007f529d863000)
	libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007f529d85d000)
	libkrb5support.so.0 => /usr/lib64/libkrb5support.so.0 (0x00007f529d84d000)
	libkeyutils.so.1 => /usr/lib64/libkeyutils.so.1 (0x00007f529d847000)
	libresolv.so.2 => /lib64/libresolv.so.2 (0x00007f529be2a000)
	libselinux.so.1 => /lib64/libselinux.so.1 (0x00007f529bc01000)
	libpcre.so.1 => /usr/lib64/libpcre.so.1 (0x00007f529bb71000)
matej@stitny: tmp$
```
I guess, the answer is yes. Note however, that I have actually never run binary out of this package myself, so I would have to find some Fedora machine to test it (and aside from openSUSE machines, I have only some CentOS7 ones here).

jamessan commented:
> Is Nvim being _linked to_ Lua 5.2? That's a bit surprising since I thought the 5.1/5.2 C APIs were incompatible.

The different Lua versions aren't _explicitly_ API incompatible.  There simply are no guarantees about compatibility, so depending on the APIs being used it is possible for one code base to work in all the Lua versions.

Then there's the matter that different Lua versions definitely are _not_ ABI compatible, so it's not possible to use compiled C modules that were built against different Lua versions.

mcepl commented:
> Then there's the matter that different Lua versions definitely are _not_ ABI compatible, so it's not possible to use compiled C modules that were built against different Lua versions.

Well, in the world of Linux distributions, you are expected to have rebulit all binary modules from its sources, so it shouldn't be a problem.

mcepl commented:
> @mcepl I pushed to your branch. Could you confirm if this works? Then will merge.

It works, but I have a comment.

justinmk committed:
lint

justinmk commented on runtime/lua/vim/compat.lua:
```diff
@@ -0,0 +1,12 @@
+
+local lua_version = _VERSION:sub(-3)
+
+if lua_version >= "5.2" then
```
Cute trick from [lua-compat-5.3](https://github.com/keplerproject/lua-compat-5.3).

mcepl replied:
Yes, that's what I meant. Thank you.

justinmk commented:
@mcepl see latest push .
